### ISSUE TITLE
[Autograd]AccumulateGrad Node runs on a different stream when model has specific stream

### ISSUE THREAD

**Issue by majing921201** [pytorch/pytorch#137483]
hi, We found that in follow case, AccumulateGrad runs on a different stream with other nodes
```python
import torch

N, D_in, H, D_out = 640, 4096, 2048, 1024
model = torch.nn.Sequential(torch.nn.Linear(D_in, H),
                            torch.nn.Dropout(p=0.2),
                            torch.nn.Linear(H, D_out),
                            torch.nn.Dropout(p=0.1)).cuda()
loss_fn = torch.nn.MSELoss()
optimizer = torch.optim.SGD(model.parameters(), lr=0.1)

input = torch.randn(N, D_in, device='cuda')
target = torch.randn(N, D_out, device='cuda')

# 1, warmup stage
s = torch.cuda.Stream()                                           # s is stream3
s.wait_stream(torch.cuda.current_stream())                        # all nodes runs on stream3
with torch.cuda.stream(s):
    for i in range(3):
        optimizer.zero_grad(set_to_none=True)
        y_pred = model(input)
        loss = loss_fn(y_pred, target)
        loss.backward()
        optimizer.step()
torch.cuda.current_stream().wait_stream(s)

# 2, capture stage
optimizer.zero_grad(set_to_none=True)             # current stream is 0
y_pred = model(input)                             # all nodes runs on stream0, except AccumulateGrad still on stream3              
loss = loss_fn(y_pred, target)
loss.backward()
```
This usage is extracted from cuda graph, but I think it is also a general usage if user wants to switch streams in different runs.
1, warmup stage, runs model on one stream (stream3)
2, capture stage, runs model on current stream (stream0)

We found that in the 2nd stage, AccumulateGrad node still runs on stream3. Root cause is that each backward CUDA op runs on the same stream that was used for its corresponding forward op [Stream semantics of backward passes](https://pytorch.org/docs/stable/notes/cuda.html#stream-semantics-of-backward-passes). Stream info for each op is recorded in node, which is created in forward pass. AccumulateGrad node is a little special, who is reused when create [grad_accumulator](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/variable.cpp#L301) . `grad_accumulator_` is a member of AutogradMeta. [AutogradMeta](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/variable.h#L200) is one unique struct of `Variable`.
So in this case, Variable likes Linear weight doesn't change, accumulator wouldn't change too. When we run model first time, stream3 is recorded for AccumulateGrad node, and wouldn't change later. And when backward() called, Autograd engine makes AccumulateGrad node always run on stream3. Although we use specific steam0 in the second stage.
For my understanding, user has assigned second stage runs on current stream (stream0). That means all of the model nodes should be on stream0. No node should run on any other stream. please help check this usage. Thanks.

cc @ezyang @albanD @gqchen @pearu @nikitaved @soulitzer @Varal7 @xmfan @ptrblck @msaroufim

**Comment by soulitzer** [comment]
This is expected. The grad_accumulator node (which was created under stream3) is being reused between the two forwards because it is being kept by the autograd graph, which is kept alive by the output tensor. 

If you delete the output/loss of the first forward before running the second forward, the accumulate grad backward would be run on stream0.

**Comment by majing921201** [comment]
Thanks for your explanation, @soulitzer. Can you show me how to delete the loss in my case, after the first stage finished ?
I tried `del loss` but doesn't work.

**Comment by soulitzer** [comment]
Just to double check, are you also doing `del y_pred`?

**Comment by soulitzer** [comment]
Closing as this is expected, but feel free to reach out if you are still running into issues @majing921201.
